lib/content_encoding.c

```diff
@@ -132,65 +133,85 @@ inflate_stream(struct connectdata *conn, contenc_writer *writer)
   /* Dynamically allocate a buffer for decompression because it's uncommonly
      large to hold on the stack */
   decomp = malloc(DSIZ);
-  if(decomp == NULL) {
+  if(decomp == NULL)
     return exit_zlib(conn, z, &zp->zlib_init, CURLE_OUT_OF_MEMORY);
-  }
 
   /* because the buffer size is fixed, iteratively decompress and transfer to
-     the client via client_write. */
-  for(;;) {
-    if(z->avail_in == 0) {
-      free(decomp);
-      return result;
+     the client via downstream_write function. */
+  while(!result && z->avail_in) {
+    /* Check state. */
+    switch(zp->zlib_init) {
+    case ZLIB_INIT:
+    case ZLIB_INFLATING:
+    case ZLIB_INIT_GZIP:
+    case ZLIB_GZIP_INFLATING:
+      break;
+    default:
+      result = exit_zlib(conn, z, &zp->zlib_init, CURLE_WRITE_ERROR);
+      continue;
```

Comment:
I think this would be easier to understand:
~~~c
if(zp->zlib_init != ZLIB_INIT &&
   zp->zlib_init != ZLIB_INFLATING &&
   zp->zlib_init != ZLIB_INIT_GZIP &&
   zp->zlib_init != ZLIB_GZIP_INFLATING) {
  result = exit_zlib(conn, z, &zp->zlib_init, CURLE_WRITE_ERROR);
  break;
}
~~~

Reply:
Many thanks for the review.
Maybe you're right. At least we have the same effect.